src/client/client-wrapper.ts

```diff
@@ -42,25 +61,76 @@ export class ClientWrapper {
    *   simplify automated testing. Should default to the class/constructor of
    *   the underlying/wrapped API client.
    */
-  constructor (auth: grpc.Metadata, clientConstructor = needle) {
-    // Call auth.get() for any field defined in the static expectedAuthFields
-    // array here. The argument passed to get() should match the "field" prop
-    // declared on the definition object above.
-    const uaString: string = auth.get('userAgent').toString();
-    this.client = clientConstructor;
+  // tslint:disable-next-line:max-line-length
+  constructor(auth: grpc.Metadata, clientConstructor = DynamicsWebApi, adal = AuthenticationContext) {
+    const authContext = adal.AuthenticationContext;
+    const adalContext = new authContext(`https://login.microsoftonline.com/${auth.get('tenantId')[0]}/oauth2/token`);
 
-    // Authenticate the underlying client here.
-    this.client.defaults({ user_agent: uaString });
+    function acquireToken(dynamicsWebApiCallback) {
+      function adalCallback(error, token) {
+        if (!error) {
+          dynamicsWebApiCallback(token);
```

Comment:
- [ ] To prevent flakiness / race conditions, do we need to set `this.clientReady = Promise.resolve(true)` here?